vm/src/vm.h

```diff
@@ -14,10 +15,43 @@
 extern "C" {
 #endif
 
+typedef union {
+    int64_t i;
+    uint8_t b;
+    double d;
+    const void *ptr;
+} moshell_value;
+
+typedef enum {
+    OBJ_STR,
+    OBJ_INT,
+    OBJ_DOUBLE,
+    OBJ_VEC
+} moshell_object_type;
+
+typedef struct {
+    moshell_object_type type;
+    const void *val;
+} moshell_object;
+
+typedef struct {
+    const uint64_t size;
+    const moshell_value *data;
+} moshell_array;
+
+uint8_t moshell_value_get_as_byte(moshell_value val);
+int64_t moshell_value_get_as_int(moshell_value val);
+double moshell_value_get_as_double(moshell_value val);
+moshell_object moshell_value_get_as_object(moshell_value val);
+
+moshell_value moshell_object_unbox(moshell_object obj);
+const char *moshell_object_get_as_string(moshell_object obj);
+moshell_array moshell_object_get_as_array(moshell_object obj);
```

Comment:
Please document public functions, and move them below, `moshell_exec/init/register/run` are the main method of this file.